<a href="https://colab.research.google.com/github/drshahizan/Python-big-data/blob/main/Project/SIX/Final_project_hpdp.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **NYC Parking Ticket Analysis Using 3 Different Libraries.**

**Group Members:**

<table>
  <tr>
    <th>Name</th>
    <th>Matric</th>
  </tr>
  <tr>
    <th>LEE MING QI</th>
    <th>A20EC0064</th>
  </tr>

  <tr>
    <th>AMIRAH RAIHANAH BINTI ABDUL RAHIM</th>
    <th>A20EC0182</th>
  </tr>
    <tr>
    <th>SINGTHAI SRISOI</th>
    <th>A20EC0147</th>
  </tr>
    <tr>
    <th>NUR IRDINA ALIAH BINTI ABDUL WAHAB</th>
    <th>A20EC0115</th>
  </tr>
</table>


**About Project:**

In this project, we will analyze NYC Parking Ticket dataset by performing EDA using 3 different libraries. The libraries that will be use are Dask, Pyspark and Koalas.The dataset used can be retrieve from [Kaggle](https://www.kaggle.com/datasets/new-york-city/nyc-parking-tickets). This is a large dataset that contains 10 million rows and 51 columns with the size of 2GB. We will show each operation by using the 3 libraries stated and provide time taken needed to perform it. This will help you to evaluate which library can do it more effectively. 
This project is divided to 4 parts;


1.   Downloading Dataset
2.   Data Cleaning
3.   Data Visualization
4.   Data Questions and Answers
5.   Findings and Analysis






---


## **`Downloading the Dataset`**

The NYC Department of Finance collects data on every parking ticket issued in NYC. This data is made publicly available to aid in ticket resolution and to guide policymakers. Now, we will perform EDA by first downloading the dataset from Kaggle [here](https://www.kaggle.com/datasets/new-york-city/nyc-parking-tickets).

Download steps:
1. Open [dataset link](https://www.kaggle.com/datasets/new-york-city/nyc-parking-tickets) that will be directed to the NYC Parking Ticket dataset.
2. Download the dataset.
3. Upload dataset into Google Drive.
4. Get link of the dataset file in Google Drive.
5. Import the dataset retrieve from Google Drive.

In [1]:
# Pull dataset from google drive
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
link = 'https://drive.google.com/file/d/1TSEOGseOzszVn342YESVHk7LyjX1Qmbd/view?usp=sharing'
# to get the id part of the file
id = link.split("/")[-2]
 
downloaded = drive.CreateFile({'id':id})
downloaded.GetContentFile('NYC_Parking_Ticket.csv') 



---

# **`Read Dataset Into Dataframe Of Each Library`**
Steps:
1. Install library using pip.
2. Import the library installed.
3. Read the csv file into dataframe.

In [2]:
import pandas as pd
import numpy as np

**1. Dask**

In [3]:
!python -m pip install "dask[complete]" # Install everything

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [4]:
 #specify dtypes to prevent errors because the dataset is messy.
 dtypes={'Date First Observed': str,
    'Days Parking In Effect    ': str,
    'Double Parking Violation': str,
    'Feet From Curb': np.float32,
    'From Hours In Effect': str,
    'House Number': str,
    'Hydrant Violation': str,
    'Intersecting Street': str,
    'Issue Date': str,
    'Issuer Code': np.float32,
    'Issuer Command': str,
    'Issuer Precinct': np.float32,
    'Issuer Squad': str,
    'Issuing Agency': str,
    'Law Section': np.float32,
    'Meter Number': str,
    'No Standing or Stopping Violation': str,
    'Plate ID': str,
    'Plate Type': str,
    'Registration State': str,
    'Street Code1': np.int32,
    'Street Code2': np.int32,
    'Street Code3': np.int32,
    'Street Name': str,
    'Sub Division': str,
    'Summons Number': np.int32,
    'Time First Observed': str,
    'To Hours In Effect': str,
    'Unregistered Vehicle?': str,
    'Vehicle Body Type': str,
    'Vehicle Color': str,
    'Vehicle Expiration Date': str,
    'Vehicle Make': str,
    'Vehicle Year': np.float32,
    'Violation Code': np.int16,
    'Violation County': str,
    'Violation Description': str,
    'Violation In Front Of Or Opposite': str,
    'Violation Legal Code': str,
    'Violation Location': str,
    'Violation Post Code': str,
    'Violation Precinct': np.float32,
    'Violation Time': str
}

In [5]:
import dask.dataframe as dd

In [6]:
%%time
dfdask = dd.read_csv('NYC_Parking_Ticket.csv',dtype=dtypes)
dfdask.head()
       

CPU times: user 1.55 s, sys: 178 ms, total: 1.73 s
Wall time: 1.76 s


Summons Number Plate ID Registration State Plate Type  Issue Date  \
0      1283294138  GBB9093                 NY        PAS  08/04/2013   
1      1283294151  62416MB                 NY        COM  08/04/2013   
2      1283294163  78755JZ                 NY        COM  08/05/2013   
3      1283294175  63009MA                 NY        COM  08/05/2013   
4      1283294187  91648MC                 NY        COM  08/08/2013   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Hydrant Violation  Double Parking Violation Latitude Longitude  \
0  ...                NaN                       NaN      NaN       NaN   
1  ...                NaN                       NaN      NaN       NaN   
2  ...                NaN                       NaN      NaN       NaN   
3  ...                NaN                       NaN      NaN       NaN   
4  ...                NaN                       NaN      NaN       NaN   

   Community Board  Community Council   Census Tract BIN BBL NTA  
0              NaN                 NaN           NaN NaN NaN NaN  
1              NaN                 NaN           NaN NaN NaN NaN  
2              NaN                 NaN           NaN NaN NaN NaN  
3              NaN                 NaN           NaN NaN NaN NaN  
4              NaN                 NaN           NaN NaN NaN NaN  

[5 rows x 51 columns]

**2. PySpark**

In [7]:
!pip install pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 281.4/281.4 MB 4.7 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.7/199.7 KB 18.4 MB/s eta 0:00:00
  Created wheel for pyspark: filename=pyspark-3.3.1-py2.py3-none-any.whl size=281845512 sha256=bc27d2c7a765f22113ac83abc307d60906ae26fae15e88ff4b501a95e68f11fd
  Stored in directory: /root/.cache/pip/wheels/43/dc/11/ec201cd671da62fa9c5cc77078235e40722170ceba231d7598
Successfully built pyspark


In [8]:
from pyspark.sql import SparkSession
import pyspark.sql.functions as func
from pyspark.sql.types import StructType,StructField, StringType, IntegerType 
from pyspark.sql.types import ArrayType, DoubleType, BooleanType
from pyspark.sql.functions import col,array_contains
spark = SparkSession.builder.getOrCreate()

In [9]:
%%time
dfs = spark.read.options(header = "true", \
  inferSchema = "true",) \
     .csv("NYC_Parking_Ticket.csv")
dfs.show(5)

+--------------+--------+------------------+----------+----------+--------------+-----------------+------------+--------------+------------+------------+------------+-----------------------+------------------+------------------+---------------+-----------+--------------+------------+--------------+-------------------+----------------+---------------------------------+------------+-----------+-------------------+-------------------+-----------+------------+--------------------+--------------------------+--------------------+------------------+-------------+---------------------+------------+------------+--------------+-------------------+---------------------+---------------------------------+-----------------+------------------------+--------+---------+---------------+------------------+------------+----+----+----+
|Summons Number|Plate ID|Registration State|Plate Type|Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Issuing Agency|Street Code1|Street Code2|Street Code3|Vehicle

**3. Koalas**

In [10]:
!pip install koalas

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 390.8/390.8 KB 20.7 MB/s eta 0:00:00


In [11]:
import databricks.koalas as ks

In [12]:
%%time
dfk = ks.read_csv('NYC_Parking_Ticket.csv',dtype=dtypes)
dfk.head()

CPU times: user 4.09 s, sys: 668 ms, total: 4.75 s
Wall time: 1min 24s


Summons Number Plate ID Registration State Plate Type  Issue Date  Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  Street Code2  Street Code3 Vehicle Expiration Date Violation Location  Violation Precinct  Issuer Precinct  Issuer Code Issuer Command Issuer Squad Violation Time Time First Observed Violation County Violation In Front Of Or Opposite House Number Street Name Intersecting Street Date First Observed  Law Section Sub Division Violation Legal Code Days Parking In Effect     From Hours In Effect To Hours In Effect Vehicle Color Unregistered Vehicle?  Vehicle Year Meter Number  Feet From Curb Violation Post Code Violation Description No Standing or Stopping Violation Hydrant Violation Double Parking Violation Latitude Longitude Community Board Community Council  Census Tract   BIN   BBL   NTA
0      1283294138  GBB9093                 NY        PAS  08/04/2013              46              SUBN         AUDI              P         37250         13610         21190                20140831                 33                33.0             33.0     921043.0           0033         0000          0752A                None             None                                 F          712    W 175 ST                None                   0        408.0           F1                 None                    BBBBBBB                  ALL                ALL            GY                     0        2013.0            -             0.0                None                  None                              None              None                     None     None      None            None               None         None  None  None  None
1      1283294151  62416MB                 NY        COM  08/04/2013              46               VAN         FORD              P         37290         40404         40404                20140430                 33                33.0             33.0     921043.0           0033         0000          1240P                None               NY                                 O          201    W 177 ST                None                   0        408.0            C                 None                    BBBBBBB                  ALL                ALL            WH                     0        2012.0            -             0.0                None                  None                              None              None                     None     None      None            None               None         None  None  None  None
2      1283294163  78755JZ                 NY        COM  08/05/2013              46               P-U        CHEVR              P         37030         31190         13610                20140228                 33                33.0             33.0     921043.0           0033         0000          1243P                None               NY                                 O          520    W 163 ST                None                   0        408.0           F7                 None                    BBBBBBB                  ALL                ALL          None                     0           0.0            -             0.0                None                  None                              None              None                     None     None      None            None               None         None  None  None  None
3      1283294175  63009MA                 NY        COM  08/05/2013              46               VAN         FORD              P         37270         11710         12010                20141031                 33                33.0             33.0     921043.0           0033         0000          0232P                None               NY                                 O          517    W 176 ST                None                   0        408.0           F1                 None                    BBBBBBB                  ALL                ALL            WH                     0        2010.0            -             0

>**Findings ▶** 
All library has different way of importing the dataset and requires different functions to run it.
1. Dask 🕔
Dask is the fastest way to read the data by taking **1.79s** but we have to initialize its dtype. This is because there are a lot of mismatched data type in the dataset. It requires to import **dask.dataframe** in order to read it in dataframe. The function used is common like Pandas which is **read.csv()**.

>2. PySpark 🕔
Pyspark took a long time to read the dataset which is **1min 10s**. In pyspark, there is no need to initialize dtype because we can set **inferSchema = True** and it will assign the correct dtype. The function used is different from common Pandas which is **spark.read.options.**

>3. Koalas 🕔
Koalas took the **longest time** compared to Dask and PySpark which is **1min 10s**.
Most of Koalas funtions are the same with common Pandas which in this case we use **read.csv()** to read the data into dataframe.




---



---



# **`Data Preparation and Cleaning`**

The dataset contains 10 million rows and 51 columns of records which has a lot of missing data, mismatched data, unused columns and other mess that need to be clean. Before we can analyze the data, we must clean the data to obtain an accurate analysis. Now, we will perform data cleaning using 3 different libraries and evaluate which library does it more effectively.

## **`Step 1: Checking dataframe size to ensure all libraries read the same number of records.`**

**1. Dask**

In [13]:
%%time
dfdask.shape

CPU times: user 8.2 ms, sys: 0 ns, total: 8.2 ms
Wall time: 8.28 ms


(Delayed('int-5a52d29a-01b3-4373-a665-e7a494ecf4ab'), 51)

**2. PySpark**

In [14]:
%%time
rows = dfs.count()
cols = len(dfs.columns)
print (rows, "," ,cols)


9100278 , 51
CPU times: user 49.5 ms, sys: 3.35 ms, total: 52.8 ms
Wall time: 6.77 s


**3. Koalas**

In [15]:
%%time
dfk.shape

CPU times: user 56.3 ms, sys: 4.69 ms, total: 61 ms
Wall time: 9.66 s


(9100278, 51)

>**Findings ▶** 
All libraries managed to read the same no of rows and columns which are **9100278 rows and 51 columns**. To find the number and columns read by all libraries, **Dask and Koalas** use the same method which is **dataframe.shape**. Meanwhile, **Pyspark** is different and it **uses count to find rows and cols for columns**.

>1. Dask 🕔
Dask is the fastest way to find number of rows and columns by taking **9.06ms**.

>2. PySpark 🕔
PySpark comes next as it took **6.47s** to find number of rows and columns.

>3. Koalas 🕔
Koalas took the **longest time** compared to Dask and PySpark which is **9.49s**.




---



---



## **`Step 2: Check and drop column that has more than 80% missing values.`**



This dataset contains a lot of missing and null values and certain columns have no values at all. Anyhow, dropping all missing values is not recommended as it will lose a lot of records. Some null values can be change to 0, for example column `Feet From Curb` because the attribute may not be collected but the parking ticket is still valid.In order to get a clean data, we must drop the columns and change the null values to get consistent result for our analysis.

Steps:
1. Check number of Nan values.
2. Drop columns that have more than 80% Nan values.
3. Fill in the rest of Nan values with 0.
4. Check if there is any Nan values left.

**1. Dask**

In [16]:
#check number of nan values
%%time
dfdask.isnull().sum().compute()


CPU times: user 1min 13s, sys: 2.55 s, total: 1min 15s
Wall time: 1min 3s


Summons Number                             0
Plate ID                                3612
Registration State                         0
Plate Type                                 0
Issue Date                                 0
Violation Code                             0
Vehicle Body Type                     111377
Vehicle Make                           63273
Issuing Agency                             0
Street Code1                               0
Street Code2                               0
Street Code3                               0
Vehicle Expiration Date                    0
Violation Location                    671170
Violation Precinct                         0
Issuer Precinct                            0
Issuer Code                                0
Issuer Command                        654437
Issuer Squad                          654437
Violation Time                          2074
Time First Observed                  8080301
Violation County                      705205
Violation 

In [17]:
#drop columns that have more than 80% values
%%time
dfdask=dfdask.drop(['Time First Observed','Violation Legal Code','Unregistered Vehicle?','No Standing or Stopping Violation', 'Hydrant Violation', 'Double Parking Violation','Latitude','Longitude','Community Board','Community Council ','Census Tract','BIN','BBL','NTA'], axis=1) 
dfdask.head()


CPU times: user 1.48 s, sys: 72.3 ms, total: 1.55 s
Wall time: 1.56 s


Summons Number Plate ID Registration State Plate Type  Issue Date  \
0      1283294138  GBB9093                 NY        PAS  08/04/2013   
1      1283294151  62416MB                 NY        COM  08/04/2013   
2      1283294163  78755JZ                 NY        COM  08/05/2013   
3      1283294175  63009MA                 NY        COM  08/05/2013   
4      1283294187  91648MC                 NY        COM  08/08/2013   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Sub Division  Days Parking In Effect     From Hours In Effect  \
0  ...            F1                     BBBBBBB                  ALL   
1  ...             C                     BBBBBBB                  ALL   
2  ...            F7                     BBBBBBB                  ALL   
3  ...            F1                     BBBBBBB                  ALL   
4  ...            E1                     BBBBBBB                  ALL   

  To Hours In Effect  Vehicle Color  Vehicle Year  Meter Number  \
0                ALL             GY        2013.0             -   
1                ALL             WH        2012.0             -   
2                ALL            NaN           0.0             -   
3                ALL             WH        2010.0             -   
4                ALL             BR        2012.0             -   

  Feet From Curb Violation Post Code Violation Description  
0            0.0                 NaN                   NaN  
1            0.0                 NaN                   NaN  
2            0.0                 NaN                   NaN  
3            0.0                 NaN                   NaN  
4            0.0                 NaN                   NaN  

[5 rows x 37 columns]

In [18]:
#Not all nan values need to be drop. Hence we fill the Nan values with 0.
%%time
dfdask = dfdask.fillna(0)


CPU times: user 13.1 ms, sys: 0 ns, total: 13.1 ms
Wall time: 14.3 ms


In [19]:
#check again all null values to ensure it is clean.
dfdask.isnull().sum().compute()

Summons Number                       0
Plate ID                             0
Registration State                   0
Plate Type                           0
Issue Date                           0
Violation Code                       0
Vehicle Body Type                    0
Vehicle Make                         0
Issuing Agency                       0
Street Code1                         0
Street Code2                         0
Street Code3                         0
Vehicle Expiration Date              0
Violation Location                   0
Violation Precinct                   0
Issuer Precinct                      0
Issuer Code                          0
Issuer Command                       0
Issuer Squad                         0
Violation Time                       0
Violation County                     0
Violation In Front Of Or Opposite    0
House Number                         0
Street Name                          0
Intersecting Street                  0
Date First Observed      

**2.PySpark**

In [20]:
#count missing values
%%time
from pyspark.sql.functions import isnan, when, count, col
dfs.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in dfs.columns]).show()

+--------------+--------+------------------+----------+----------+--------------+-----------------+------------+--------------+------------+------------+------------+-----------------------+------------------+------------------+---------------+-----------+--------------+------------+--------------+-------------------+----------------+---------------------------------+------------+-----------+-------------------+-------------------+-----------+------------+--------------------+--------------------------+--------------------+------------------+-------------+---------------------+------------+------------+--------------+-------------------+---------------------+---------------------------------+-----------------+------------------------+--------+---------+---------------+------------------+------------+-------+-------+-------+
|Summons Number|Plate ID|Registration State|Plate Type|Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Issuing Agency|Street Code1|Street Code2|Street Code

In [21]:
#drop columns that have more than 80% missing values
%%time
dfs.drop('Time First Observed','Violation Legal Code','Unregistered Vehicle?','No Standing or Stopping Violation', 'Hydrant Violation', 'Double Parking Violation','Latitude','Longitude','Community Board','Community Council ','Census Tract','BIN','BBL','NTA') \
    .printSchema()

root
 |-- Summons Number: long (nullable = true)
 |-- Plate ID: string (nullable = true)
 |-- Registration State: string (nullable = true)
 |-- Plate Type: string (nullable = true)
 |-- Issue Date: string (nullable = true)
 |-- Violation Code: integer (nullable = true)
 |-- Vehicle Body Type: string (nullable = true)
 |-- Vehicle Make: string (nullable = true)
 |-- Issuing Agency: string (nullable = true)
 |-- Street Code1: integer (nullable = true)
 |-- Street Code2: integer (nullable = true)
 |-- Street Code3: integer (nullable = true)
 |-- Vehicle Expiration Date: integer (nullable = true)
 |-- Violation Location: integer (nullable = true)
 |-- Violation Precinct: integer (nullable = true)
 |-- Issuer Precinct: integer (nullable = true)
 |-- Issuer Code: integer (nullable = true)
 |-- Issuer Command: string (nullable = true)
 |-- Issuer Squad: string (nullable = true)
 |-- Violation Time: string (nullable = true)
 |-- Violation County: string (nullable = true)
 |-- Violation In Fron

In [22]:
#Not all nan values need to be drop. Hence we fill the Nan values with 0.
%%time
dfs=dfs.fillna(0)

CPU times: user 2.57 ms, sys: 0 ns, total: 2.57 ms
Wall time: 30.3 ms


In [23]:
#check again to ensure the data is clean
%%time
from pyspark.sql.functions import isnan, when, count, col
dfs.select([count(when(isnan(c) | col(c).isNull(), c)).alias(c) for c in dfs.columns]).show()

+--------------+--------+------------------+----------+----------+--------------+-----------------+------------+--------------+------------+------------+------------+-----------------------+------------------+------------------+---------------+-----------+--------------+------------+--------------+-------------------+----------------+---------------------------------+------------+-----------+-------------------+-------------------+-----------+------------+--------------------+--------------------------+--------------------+------------------+-------------+---------------------+------------+------------+--------------+-------------------+---------------------+---------------------------------+-----------------+------------------------+--------+---------+---------------+------------------+------------+-------+-------+-------+
|Summons Number|Plate ID|Registration State|Plate Type|Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Issuing Agency|Street Code1|Street Code2|Street Code

**3.Koalas**

In [24]:
#count missing values
%%time
dfk.isnull().sum()

CPU times: user 1.38 s, sys: 168 ms, total: 1.55 s
Wall time: 1min 15s


Summons Number                             0
Plate ID                                   0
Registration State                         0
Plate Type                                 0
Issue Date                                 0
Violation Code                             0
Vehicle Body Type                          0
Vehicle Make                               0
Issuing Agency                             0
Street Code1                               0
Street Code2                               0
Street Code3                               0
Vehicle Expiration Date                    0
Violation Location                         0
Violation Precinct                         0
Issuer Precinct                            0
Issuer Code                                0
Issuer Command                             0
Issuer Squad                               0
Violation Time                             0
Time First Observed                        0
Violation County                           0
Violation 

In [25]:
#drop columns that have more than 80% missing values
%%time
dfk=dfk.drop(['Time First Observed','Violation Legal Code','Unregistered Vehicle?','No Standing or Stopping Violation', 'Hydrant Violation', 'Double Parking Violation','Latitude','Longitude','Community Board','Community Council ','Census Tract','BIN','BBL','NTA'], axis=1) 
dfk.head()

CPU times: user 52.4 ms, sys: 6.95 ms, total: 59.4 ms
Wall time: 159 ms


Summons Number Plate ID Registration State Plate Type  Issue Date  Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  Street Code2  Street Code3 Vehicle Expiration Date Violation Location  Violation Precinct  Issuer Precinct  Issuer Code Issuer Command Issuer Squad Violation Time Violation County Violation In Front Of Or Opposite House Number Street Name Intersecting Street Date First Observed  Law Section Sub Division Days Parking In Effect     From Hours In Effect To Hours In Effect Vehicle Color  Vehicle Year Meter Number  Feet From Curb Violation Post Code Violation Description
0      1283294138  GBB9093                 NY        PAS  08/04/2013              46              SUBN         AUDI              P         37250         13610         21190                20140831                 33                33.0             33.0     921043.0           0033         0000          0752A             None                                 F          712    W 175 ST                None                   0        408.0           F1                    BBBBBBB                  ALL                ALL            GY        2013.0            -             0.0                None                  None
1      1283294151  62416MB                 NY        COM  08/04/2013              46               VAN         FORD              P         37290         40404         40404                20140430                 33                33.0             33.0     921043.0           0033         0000          1240P               NY                                 O          201    W 177 ST                None                   0        408.0            C                    BBBBBBB                  ALL                ALL            WH        2012.0            -             0.0                None                  None
2      1283294163  78755JZ                 NY        COM  08/05/2013              46               P-U        CHEVR              P         37030         31190         13610                20140228                 33                33.0             33.0     921043.0           0033         0000          1243P               NY                                 O          520    W 163 ST                None                   0        408.0           F7                    BBBBBBB                  ALL                ALL          None           0.0            -             0.0                None                  None
3      1283294175  63009MA                 NY        COM  08/05/2013              46               VAN         FORD              P         37270         11710         12010                20141031                 33                33.0             33.0     921043.0           0033         0000          0232P               NY                                 O          517    W 176 ST                None                   0        408.0           F1                    BBBBBBB                  ALL                ALL            WH        2010.0            -             0.0                None                  None
4      1283294187  91648MC                 NY        COM  08/08/2013              41              TRLR          GMC              P         37240         12010         31190                       0                 33                33.0             33.0     921043.0           0033         0000          1239P               NY                                 F          590    W 174 ST                None                   0        408.0           E1                    BBBBBBB                  ALL                ALL            BR        2012.0            -             0.0                None                  None

In [26]:
#Not all nan values need to be drop. Hence we fill the Nan values with 0.
%%time
dfk = dfk.fillna(0)

CPU times: user 353 ms, sys: 52.7 ms, total: 406 ms
Wall time: 2.21 s


In [27]:
#check again to ensure the data is clean
dfk.isnull().sum()

Summons Number                       0
Plate ID                             0
Registration State                   0
Plate Type                           0
Issue Date                           0
Violation Code                       0
Vehicle Body Type                    0
Vehicle Make                         0
Issuing Agency                       0
Street Code1                         0
Street Code2                         0
Street Code3                         0
Vehicle Expiration Date              0
Violation Location                   0
Violation Precinct                   0
Issuer Precinct                      0
Issuer Code                          0
Issuer Command                       0
Issuer Squad                         0
Violation Time                       0
Violation County                     0
Violation In Front Of Or Opposite    0
House Number                         0
Street Name                          0
Intersecting Street                  0
Date First Observed      

>**Findings ▶** 
There are some similarity in all 3 libraries in finding and drop missing values. Dask and Koalas uses the same method but PySpark using different method.
We found that the fastest library that can do this operation is **Dask** with the total of **1min 54s**.
>1. Dask 🕔
Dask took **1min 4s** to compute nan values **1.52s** to drop columns and **14.18ms** to fill null values. Total execution time is **1min 6s**, the fastest library.

>2. PySpark 🕔
PySpark took **7min 50s** to compute nan values **30.6s** to drop columns and **39ms** to fill null values. Total execution time is **8min 34s** which is **8 times slower compared to Dask**.

>3. Koalas 🕔
PySpark took **1min 12s** to compute nan values **171ms** to drop columns and **2.33s** to fill null values. Total execution time is **1min 15s** which is **9s behind Dask**.





---



---



# **`Step 3: Convert column 'Issue Date' to datetime.`**

The ` Issue Date`column has string type data hence it need to be change to datetime type.

**1. Dask**

In [28]:
%%time
dfdask['Issue Date'] = dd.to_datetime(dfdask['Issue Date'],format='%m/%d/%Y',errors='coerce')
dfdask.dtypes

CPU times: user 19.6 ms, sys: 0 ns, total: 19.6 ms
Wall time: 29.5 ms


Summons Number                                int32
Plate ID                                     object
Registration State                           object
Plate Type                                   object
Issue Date                           datetime64[ns]
Violation Code                                int16
Vehicle Body Type                            object
Vehicle Make                                 object
Issuing Agency                               object
Street Code1                                  int32
Street Code2                                  int32
Street Code3                                  int32
Vehicle Expiration Date                      object
Violation Location                           object
Violation Precinct                          float32
Issuer Precinct                             float32
Issuer Code                                 float32
Issuer Command                               object
Issuer Squad                                 object
Violation Ti

**2. PySpark**

In [29]:
%%time
#dfs = dfs.withColumn("Issue Date", to_date(dfs.date_string))
#from pyspark.sql.functions import to_date, lit
#df.select("db_id", to_date(lit("1-12-2022"),"d-MM-yyyy")).show()
from pyspark.sql.types import StringType, DateType, FloatType
dfs = dfs.withColumn('Issue Date', func.to_date('Issue Date', 'dd/MM/yyyy'))
dfs.printSchema()


root
 |-- Summons Number: long (nullable = true)
 |-- Plate ID: string (nullable = true)
 |-- Registration State: string (nullable = true)
 |-- Plate Type: string (nullable = true)
 |-- Issue Date: date (nullable = true)
 |-- Violation Code: integer (nullable = true)
 |-- Vehicle Body Type: string (nullable = true)
 |-- Vehicle Make: string (nullable = true)
 |-- Issuing Agency: string (nullable = true)
 |-- Street Code1: integer (nullable = true)
 |-- Street Code2: integer (nullable = true)
 |-- Street Code3: integer (nullable = true)
 |-- Vehicle Expiration Date: integer (nullable = true)
 |-- Violation Location: integer (nullable = true)
 |-- Violation Precinct: integer (nullable = true)
 |-- Issuer Precinct: integer (nullable = true)
 |-- Issuer Code: integer (nullable = true)
 |-- Issuer Command: string (nullable = true)
 |-- Issuer Squad: string (nullable = true)
 |-- Violation Time: string (nullable = true)
 |-- Time First Observed: string (nullable = true)
 |-- Violation County

**3. Koalas**

In [30]:
%%time
dfk['Issue Date'] = ks.to_datetime(dfk['Issue Date'],format='%m/%d/%Y',errors='coerce')
dfk.dtypes

/usr/local/lib/python3.8/dist-packages/pyspark/sql/pandas/functions.py:394: UserWarning: In Python 3.6+ and Spark 3.0+, it is preferred to specify type hints for pandas UDF instead of specifying pandas UDF type which will be deprecated in the future releases. See SPARK-28264 for more details.
  warnings.warn(


CPU times: user 123 ms, sys: 27 ms, total: 150 ms
Wall time: 917 ms


Summons Number                            int32
Plate ID                                 object
Registration State                       object
Plate Type                               object
Issue Date                           datetime64
Violation Code                            int32
Vehicle Body Type                        object
Vehicle Make                             object
Issuing Agency                           object
Street Code1                              int32
Street Code2                              int32
Street Code3                              int32
Vehicle Expiration Date                  object
Violation Location                       object
Violation Precinct                      float32
Issuer Precinct                         float32
Issuer Code                             float32
Issuer Command                           object
Issuer Squad                             object
Violation Time                           object
Violation County                        

>**Findings ▶** 
To change dtype of a column Dask and Koalas use the same method which is **to_datetime()**. Meanwhile, PySpark need to use import function **to_date()**.
We found that the **fastest library** is **PySpark** that took **22.6ms**.
>1. Dask 🕔
Dask took **30.9ms** to change dtype from string to datetime which is **12.3s behind PySpark.**

>2. PySpark 🕔
PySpark is the **fastest** compared to Dask and Koalas with only **22.6ms** to change dtype string to date.

>3. Koalas 🕔
Koalas remain as the **slowest library** that took **1.13s** to change dtype string to datetime.





---



---



## **`Step 4: Drop Duplicates of Unique Values`**

In NYC Parking Ticket dataset, the `Summons Number` is a unique value because there can only be 1 record of each summon took. In the dataset, we found that there are duplicates of this value. Hence, we must drop the duplicated value to prevent from errors in analyzing the data.

**1. Dask**

In [31]:
%%time
dfdask.drop_duplicates(subset = ['Summons Number'], inplace = True)
dfdask.head()


CPU times: user 2.43 s, sys: 55.8 ms, total: 2.48 s
Wall time: 2.49 s


Summons Number Plate ID Registration State Plate Type Issue Date  \
0      1283294138  GBB9093                 NY        PAS 2013-08-04   
1      1283294151  62416MB                 NY        COM 2013-08-04   
2      1283294163  78755JZ                 NY        COM 2013-08-05   
3      1283294175  63009MA                 NY        COM 2013-08-05   
4      1283294187  91648MC                 NY        COM 2013-08-08   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Sub Division  Days Parking In Effect     From Hours In Effect  \
0  ...            F1                     BBBBBBB                  ALL   
1  ...             C                     BBBBBBB                  ALL   
2  ...            F7                     BBBBBBB                  ALL   
3  ...            F1                     BBBBBBB                  ALL   
4  ...            E1                     BBBBBBB                  ALL   

  To Hours In Effect  Vehicle Color  Vehicle Year  Meter Number  \
0                ALL             GY        2013.0             -   
1                ALL             WH        2012.0             -   
2                ALL              0           0.0             -   
3                ALL             WH        2010.0             -   
4                ALL             BR        2012.0             -   

  Feet From Curb Violation Post Code Violation Description  
0            0.0                   0                     0  
1            0.0                   0                     0  
2            0.0                   0                     0  
3            0.0                   0                     0  
4            0.0                   0                     0  

[5 rows x 37 columns]

**2. PySpark**

In [32]:
#dfs.drop_duplicates(subset = ['Summons Number'], inplace = True)
%%time
dfs.dropDuplicates(['Summons Number'])
dfs.show(5)

+--------------+--------+------------------+----------+----------+--------------+-----------------+------------+--------------+------------+------------+------------+-----------------------+------------------+------------------+---------------+-----------+--------------+------------+--------------+-------------------+----------------+---------------------------------+------------+-----------+-------------------+-------------------+-----------+------------+--------------------+--------------------------+--------------------+------------------+-------------+---------------------+------------+------------+--------------+-------------------+---------------------+---------------------------------+-----------------+------------------------+--------+---------+---------------+------------------+------------+----+----+----+
|Summons Number|Plate ID|Registration State|Plate Type|Issue Date|Violation Code|Vehicle Body Type|Vehicle Make|Issuing Agency|Street Code1|Street Code2|Street Code3|Vehicle

**3. Koalas**

In [33]:
%%time
dfk.drop_duplicates(subset = ['Summons Number'], inplace = True)
dfk.head()


CPU times: user 216 ms, sys: 41.4 ms, total: 257 ms
Wall time: 1.91 s


Summons Number Plate ID Registration State Plate Type Issue Date  Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  Street Code2  Street Code3 Vehicle Expiration Date Violation Location  Violation Precinct  Issuer Precinct  Issuer Code Issuer Command Issuer Squad Violation Time Violation County Violation In Front Of Or Opposite House Number      Street Name Intersecting Street Date First Observed  Law Section Sub Division Days Parking In Effect     From Hours In Effect To Hours In Effect Vehicle Color  Vehicle Year Meter Number  Feet From Curb Violation Post Code           Violation Description
84217      -1589820192   611YJS                 CT        PAS 2013-07-29              38              4DSD        HONDA              T         54580             0             0                20130088                109               109.0            109.0     355249.0           T401            P          0240P                Q                                 I            E          Main St  29ft N/of Peck Ave                   0        408.0           h1                          Y                0900A              0700P          GREY           0.0         None             0.0                03 -  38-Failure to Display Muni Rec
84222      -1589820090  7AM9586                 MD        PAS 2013-07-30              38              SUBN        HONDA              T             0         40404         40404                20131188                109               109.0            109.0     356334.0           T401            M          0615A                Q                                 I         None  Flushing Lot #1                None                   0        408.0           h1                          Y                0600A              1000P          GREY           0.0         None             0.0                 A -  38-Failure to Display Muni Rec
502091     -1589819695  24490MD                 NY        COM 2013-07-30              51               VAN        ISUZU              T          8140          8990          9140                20150228                114               114.0            114.0     356329.0           T401            O          0333P                Q                                 O        38-06          31st St                None                   0        408.0           e3                       None                 None               None            WH        2001.0         None             0.0                 D -                     51-Sidewalk
502054     -1589819475  FHV4577                 NY        PAS 2013-07-31              74              SUBN        NISSA              T         18240          8190         57790                20150316                115               115.0            115.0     357084.0           T401            E          1014A                Q                                 F        32-45          90th St                None                   0        408.0           j2                       None                 None               None            BL        2008.0         None             0.0                16 -        74-Missing Display Plate
502133     -1589819025  FXX6070                 NY        PAS 2013-08-01              20              SUBN          GMC              T         14740         58190         45490                20140515                104               104.0            104.0     357054.0           T401            K          1123A                Q                                 O        58-11         70th Ave                None                   0        408.0            d                          Y                0800A              0600P            GY        2003.0         None             0.0                62-A        20A-No Parking (Non-COM)

>**Findings ▶** 
In order to remove or drop duplicate values Dask and Koalas use the same method which is **dataframe.drop_duplicates()** while PySpark uses **dataframe.dropDuplicates()**. For this operation, PySpark is the fastest library which took **273ms**.

>1. Dask 🕔
Dask took **2.33s** to remove the duplicates value of `Summons Number` and is the second fastest after PySpark.

>2. PySpark 🕔
PySpark is the **fastest** compared to Dask and Koalas with only **273ms** to remove duplicate values of `Summons Number` .

>3. Koalas 🕔
Koalas remain as the **slowest library** that took **12.55s** to remove duplicate values of `Summons Number` .

# Exploratory Analysis and Visualization





> Instructions (delete this cell)

* Compute the mean, sum, range and other 
interesting statistics for numeric columns
* Explore distributions of numeric columns using histograms etc.
* Explore relationship between columns using scatter plots, bar charts etc.
* Make a note of interesting insights from the exploratory analysis



**Pandas**

**Dask**

**Polars**

# Asking and Answering Questions

> Instructions (delete this cell)

* Ask at least 5 interesting questions about your dataset
* Answer the questions either by computing the results using Numpy/Pandas or by plotting graphs using Matplotlib/Seaborn
* Create new columns, merge multiple dataset and perform grouping/aggregation wherever necessary
* Wherever you're using a library function from Pandas/Numpy/Matplotlib etc. explain briefly what it does

### 1. Dask

a. What is the most common plate type?

In [34]:
%%time
result = dfdask.groupby("Plate Type").agg({"Plate Type" : "count"}).compute()
result.index.name = 'Name'
result.sort_values(by="Plate Type",ascending=False,inplace=True)
result.head(5)

CPU times: user 1min 18s, sys: 1.81 s, total: 1min 20s
Wall time: 1min 4s


Plate Type
Name            
PAS      6511094
COM      1887869
OMT       235273
SRF        96764
OMS        75041

In [35]:
dfdask.head(5)

Summons Number Plate ID Registration State Plate Type Issue Date  \
0      1283294138  GBB9093                 NY        PAS 2013-08-04   
1      1283294151  62416MB                 NY        COM 2013-08-04   
2      1283294163  78755JZ                 NY        COM 2013-08-05   
3      1283294175  63009MA                 NY        COM 2013-08-05   
4      1283294187  91648MC                 NY        COM 2013-08-08   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Sub Division  Days Parking In Effect     From Hours In Effect  \
0  ...            F1                     BBBBBBB                  ALL   
1  ...             C                     BBBBBBB                  ALL   
2  ...            F7                     BBBBBBB                  ALL   
3  ...            F1                     BBBBBBB                  ALL   
4  ...            E1                     BBBBBBB                  ALL   

  To Hours In Effect  Vehicle Color  Vehicle Year  Meter Number  \
0                ALL             GY        2013.0             -   
1                ALL             WH        2012.0             -   
2                ALL              0           0.0             -   
3                ALL             WH        2010.0             -   
4                ALL             BR        2012.0             -   

  Feet From Curb Violation Post Code Violation Description  
0            0.0                   0                     0  
1            0.0                   0                     0  
2            0.0                   0                     0  
3            0.0                   0                     0  
4            0.0                   0                     0  

[5 rows x 37 columns]

b. What registeration stae has the highest rate?

In [36]:
%%time
result = dfdask.groupby("Registration State").agg({"Plate Type" : "count"}).compute()
result.index.name = 'Name'
result.sort_values(by="Plate Type",ascending=False,inplace=True)
result.head(5)

CPU times: user 1min 20s, sys: 1.63 s, total: 1min 22s
Wall time: 1min 6s


Plate Type
Name            
NY       7029804
NJ        878677
PA        225760
CT        136973
FL        111887

c. What is the most common vehicle body type?

d. What is the most common vehicle colour?

e. What is the most common state?

In [37]:
dfdask.head()

Summons Number Plate ID Registration State Plate Type Issue Date  \
0      1283294138  GBB9093                 NY        PAS 2013-08-04   
1      1283294151  62416MB                 NY        COM 2013-08-04   
2      1283294163  78755JZ                 NY        COM 2013-08-05   
3      1283294175  63009MA                 NY        COM 2013-08-05   
4      1283294187  91648MC                 NY        COM 2013-08-08   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46              SUBN         AUDI              P         37250   
1              46               VAN         FORD              P         37290   
2              46               P-U        CHEVR              P         37030   
3              46               VAN         FORD              P         37270   
4              41              TRLR          GMC              P         37240   

   ...  Sub Division  Days Parking In Effect     From Hours In Effect  \
0  ...            F1                     BBBBBBB                  ALL   
1  ...             C                     BBBBBBB                  ALL   
2  ...            F7                     BBBBBBB                  ALL   
3  ...            F1                     BBBBBBB                  ALL   
4  ...            E1                     BBBBBBB                  ALL   

  To Hours In Effect  Vehicle Color  Vehicle Year  Meter Number  \
0                ALL             GY        2013.0             -   
1                ALL             WH        2012.0             -   
2                ALL              0           0.0             -   
3                ALL             WH        2010.0             -   
4                ALL             BR        2012.0             -   

  Feet From Curb Violation Post Code Violation Description  
0            0.0                   0                     0  
1            0.0                   0                     0  
2            0.0                   0                     0  
3            0.0                   0                     0  
4            0.0                   0                     0  

[5 rows x 37 columns]

In [38]:
result = dfdask.groupby('Plate Type').agg({'Plate Type' : 'count'}).compute()


KeyboardInterrupt: ignored

In [ ]:
result.

### 2. PySpark

a. What is the most common plate type?

b. What registeration stae has the highest rate?

c. What is the most common vehicle body type?

d. What is the most common vehicle colour?

e. What is the most common state?

In [ ]:
dfs.select('*').show(5)

### 3. Koalas

a. What is the most common plate type?

In [ ]:
%%time
dfk = ks.read_csv('NYC_Parking_Ticket.csv')
dfk.head()

In [ ]:
%%time
dfk.groupby('Plate Type').agg({'Plate Type' : 'count'})

b. What registeration state has the highest rate?

In [ ]:
%%time
dfk.groupby('Registration State').count()

c. What is the most common vehicle body type?

In [ ]:
%%time
dfk.groupby('Vehicle Body Type').count()

d. What is the most common vehicle colour?

In [ ]:
%%time
dfk.groupby('').count()

e. What is the most common state?

In [ ]:
%%time
dfk.groupby('').count()

# Inferences and Conclusion

**TODO** - Write some explanation here: a summary of all the inferences drawn from the analysis, and any conclusions you may have drawn by answering various questions.

# References and Future Work

**TODO** - Write some explanation here: ideas for future projects using this dataset, and links to resources you found useful.